float32 bx
float32 by
float32 bz